### PR TITLE
69 support for regression tasks and feature selection

### LINKED ISSUE #69: Support for regression tasks and feature selection
Thank you very much for your hard work on creating a good python package like NiaAML. Could you please consider to support the regression tasks and feature selection with NiaAML? 

I use remotely-based data (satellite image) to retrieve the biophysical parameter (blue carbon ;-)) through the machine learning regression and need to select the most contributed features from a  suite of input features. 

Many thanks,
Thang

### PR DESCRIPTION
**Pull request by LaurenzBeck**
> 🙇💖 Thank you for contributing to NiaAML!

## 💬 Description

This PR adds pipeline components to solve regression tasks to NiaAML.
The additions have been implemented in such a way, that existing classification code does not have to be updated.
-> This PR represents a **minor** version bump.

## ❗ Issue Links
Closes #69 

## 🧪 How Has This Been Tested?

Unit tests have been added to

- test that the data readers can deal with continuous targets
- all the component factories work with the new components
- verify that the new classifiers run without errors
- test the overall regression pipeline optimization

# ✅ Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: niaaml/classifiers/__init__.py [hunk 1]
```diff
@@ -23,9 +28,14 @@
     "Bagging",
     "ExtremelyRandomizedTrees",
     "DecisionTree",
+    "DecisionTreeRegression",
```
Comment (LaurenzBeck) [inline niaaml/classifiers/__init__.py:31]: I opted for the post-fix "Regression" to indicate if a classifier is a regressor.

File: niaaml/pipeline.py [hunk 2]
```diff
@@ -461,34 +462,40 @@ def evaluate_pipeline(
         ]
         solution_index = 0
 
-        for i in params_all:
+        # iterate over parameters and components and pass the prepared parameters to `set_parameters`
+        for params, component in params_all:
```
Comment (LaurenzBeck) [inline niaaml/pipeline.py:466]: this small refactoring makes the code block easier to follow

File: niaaml/pipeline.py [hunk 3]
```diff
@@ -509,10 +516,15 @@ def evaluate_pipeline(
...
+
+        # infer task for the selection of the OptimizationStats
+        # only regression models have the "Task" attribute
+        classification = not hasattr(classifier, 'Task')
+
         return (
             fitness_function.get_fitness(predictions, y_test) * -1,
             selected_features_mask,
-            OptimizationStats(predictions, y_test),
+            OptimizationStats(predictions, y_test) if classification else RegressionOptimizationStats(predictions, y_test),
```
Comment (LaurenzBeck) [inline niaaml/pipeline.py:527]: Even though a more general `OptimizationStats` class would have been prettier and better to maintain in the future, implementing it this way is less invasive and makes sure that other code depending on this class still works.

**Comment by LaurenzBeck** [comment]
I do not really get why the "NiaAML / build (macos-latest, 3.10) (pull_request)" action fails.

I had a look at the details and it is using `poetry run pytest`, even though the [developer guide](https://github.com/firefly-cpp/NiaAML/blob/master/CONTRIBUTING.md#testing) suggests using `poetry run coverage run --source=niaaml -m unittest discover -b`

I tried both locally and they both pass.

Do you know what I can do to make it pass in the github actions @firefly-cpp ?

File: niaaml/classifiers/__init__.py (hunk 1 above)
Comment (firefly-cpp) [reply niaaml/classifiers/__init__.py:31]: Perfect.

File: niaaml/data/csv_data_reader.py [hunk 4]
```diff
@@ -25,17 +25,19 @@ class CSVDataReader(DataReader):
         * :class:`niaaml.data.DataReader`
     """
 
-    def _set_parameters(self, src, contains_classes=True, has_header=False, **kwargs):
+    def _set_parameters(self, src, contains_classes=True, has_header=False, ignore_columns=[], **kwargs):
```
[the remark this answers is not in the document]
Comment (firefly-cpp) [reply niaaml/data/csv_data_reader.py:28]: Good point!

File: niaaml/fitness/mse.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            expected (pandas.core.series.Series): Expected values.
+
+        Returns:
+            float: Calculated fitness value.
+        """
+        return  - mean_squared_error(expected, predicted)
+    
+    def get_bounds(self):
+        #! float("-inf") leads to errors in the pipeline logic, so we use a very big number instead
+        return (-1000000000.0, 0.0)
```
[the remark this answers is not in the document]
Comment (firefly-cpp) [reply niaaml/fitness/mse.py:41]: OK.

**Review by firefly-cpp — APPROVED** [review]
@LaurenzBeck, this is amazing. One complex functionality in such a short time. Wonderful.

I do not have any significant comments, but I will test it soon. Are we releasing a new version of this package so we can fix it quickly?

Please also add detailed instructions in the documentation about adding a new regressor to this package.

**Comment by firefly-cpp** [comment]
I am still trying to figure out the macos problem. However, we may exclude 3.10 version of Python. What do you think?

**Comment by LaurenzBeck** [comment]
> Please also add detailed instructions in the documentation about adding a new regressor to this package.

I added an example section and instructions in the `README.md`.

Do you want me to add the same passage to the sphynx documentation in `docs/`?

> I am still trying to figure out the macos problem. However, we may exclude 3.10 version of Python. What do you think?

I don't think the problem is with the python version, but I may be wrong there.
In my opinion supporting >=3.11 should be fine, as we are already approaching the 3.13 release :)

**Comment by firefly-cpp** [comment]
Please update also sphinx. Definitely, let's go for > 3.10.

**Pull request merged**